phase_3/let-s-practise-custom-view.md

```diff
@@ -0,0 +1,211 @@
+
+从需求中我们可以知道：
+- 该控件需要展示图片
+- 该控件需要按照规则生成图像
+- 一般头像都是圆形
+
+![](http://ww3.sinaimg.cn/large/801b780agw1f7ugm27t4oj205m05agll.jpg)
+
+大致上可以知道是这样的。
+开搞！
+
+###  继承 ImageView 开始
+
+我们都知道 Android 自带了很多控件，我们自定义控件的出发点只是官方提供的控件无法满足业务需求的时候。
+从我们的需求来看，该控件是图片展示类的，所以我们很自然想到了只需要在系统 ImageView 上进行功能拓展即可，这样就可以满足新的需求又不会失去 ImageView 自带的功能。
+
+```Java
+public class CharAvatarView extends ImageView {
+    private static final String TAG = CharAvatarView.class.getSimpleName();
+    // 颜色画板集
+    private static final String[] colors = {
+        "#1abc9c", "#16a085", "#f1c40f", "#f39c12", "#2ecc71",
+        "#27ae60", "#e67e22", "#d35400", "#3498db", "#2980b9",
+        "#e74c3c", "#c0392b", "#9b59b6", "#8e44ad", "#bdc3c7",
+        "#34495e", "#2c3e50", "#95a5a6", "#7f8c8d", "#ec87bf",
+        "#d870ad", "#f69785", "#9ba37e", "#b49255", "#b49255", "#a94136"
+    };
+
+    private Paint mPaint;
+
+    private Rect mRect;
+
+    private String text;
+
+    private int charHash;
+
+    public CharAvatarView(Context context) {
+        super(context);
+    }
+
+    public CharAvatarView(Context context, AttributeSet attrs) {
+        super(context, attrs);
+        mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
+        mRect = new Rect();
+    }
+}
+```
+在这里我做了一些初始化工作，并且在其中的一个构造函数中实例化了 `Paint` 和 `Rect` 。
+
+关于 View 的构造函数的区别：
+
+```Java
+public CharAvatarView(Context context) {
+    super(context);
+}
+public CharAvatarView(Context context, AttributeSet attrs) {
+    super(context, attrs);
+}
+public CharAvatarView(Context context, AttributeSet attrs, int defStyleAttr) {
+    super(context, attrs, defStyleAttr);
+}
+```
+- 第一种属于程序内实例化时采用，之传入 Context 即可
+```Java
+CharAvatarView avatarView = new CharAvatarView(this);
+```
+这样我们的 View 就新建出来了，根据需求添加到布局即可。
+
+- 第二种用于 layout 文件实例化，会把 XML 内的参数通过 AttributeSet 带入到 View 内。
+
+- 第三个主题的 style 信息，也会从 XML 里带入
+
+目前我们的 CharAvatarView 不需要自定义主题配置，所以一般用第二种就行。
+
+### 工作流程
+我们的 View 系统是如何将它绘制到屏幕上的呢？
+
+> View 的绘制流程是从 ViewRoot 的 `performTraversals` 方法开始，它经过 measure 、 layout 和 draw 三个过程才能最终将一个 View 绘制出来，其中 measure 用来测量 View 的宽和高，layout 用来确定 View 在父容器中的放置位置，而 draw 则负责将 View 绘制在屏幕上。针对 performTraversals 的大致流程如图：
+
+![](http://ww4.sinaimg.cn/large/801b780agw1f7um4igvkcj20up0imta0.jpg)
+
+> Measure 过程决定了 View 的宽/高， Measure 完成以后，可以通过 `getMeasuredWidth` 和 `getMeasuredHeight` 方法来获取到 View 测量后的宽/高，在几乎所有的情况下它都等同于 View 最终的宽/高，但是特殊情况除外。
+> Layout 过程 决定了 View 的四个顶点的坐标和实际的 View 的宽/高，完成以后，可以通过 `getTop`、`getBottom`、`getLeft`、`getRight` 来拿到 View 的四个顶点的位置，并可以通过 `getWidth` 和 `getHeight` 方法拿到 View 最终的宽/高。
+> Draw 过程则决定了 View 的显示，只有 draw 方法完成以后 View 的内容才能呈现在屏幕上。
+
+关于 View 工作流程的深入我们在以后另外开篇进行研究。目前我们已经从宏观了解到了 View 会经历三个过程绘制出来，而且清楚了其中不同方法中的用途。接下来我们看看 CharAvatarView 在这三个流程中分别做了什么。
+
+#### onMeasure()
+
+```Java
+@Override
+protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
+    super.onMeasure(widthMeasureSpec, widthMeasureSpec); // 宽高相同
+}
+```
+让宽高相同，我在这里是只直接传入宽度进行测量。
+这样会得到一个正方形的 View。
+
+x
```

Comment:
what's this?

![](http://d.ifengimg.com/mw604/y0.ifengimg.com/ifengimcp/pic/20160418/5628dd6ecd9fa100f371_size30_w521_h534.jpg)